File: 0958-check-completeness-of-a-binary-tree/0958-check-completeness-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean solve(TreeNode root, int i, int node){
        if(root == null)
            return true;
        if(i >= node)
            return false;
        else{
            boolean left = solve(root.left, 2*i+1, node);
            boolean right = solve(root.right, 2*i+2, node);
            return (left && right);
        }
    }
    public int countNodes(TreeNode root){
        if(root == null)
            return 0;
        return 1 + countNodes(root.left) + countNodes(root.right);
    }
    public boolean isCompleteTree(TreeNode root) {
        int node = countNodes(root);
        return solve(root, 0, node);
    }
}

File: 0958-check-completeness-of-a-binary-tree/0958-check-completeness-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCompleteTree(TreeNode* root) {
        queue<pair<TreeNode*, int>> q;
        q.push({root, 1});
        int count = 1;
        while(!q.empty()){
            TreeNode* node = q.front().first;
            int value = q.front().second;
            q.pop();
            if(value > count)
                return false;
            if(node->left)
                q.push({node->left, 2*value});
            if(node->right)
                q.push({node->right, 2*value + 1});
            count++;
        }
        return true;
    }
};